### PR TITLE
Remove code duplication and use more lightweight method finding user's fields

### PR DESCRIPTION
**Pull request by hazzik**

### CODE REVIEW

File: src/NHibernate/Util/TypeExtensions.cs [hunk 1]
```diff
@@ -48,5 +52,28 @@ internal static System.Type UnwrapIfNullable(this System.Type type)
...
+
+					analyzing = analyzing.BaseType;
+				}
+			}
+		}
+
+		public static IEnumerable<FieldInfo> GetUserDeclaredFields(System.Type type)
+		{
+			// can't find another way to exclude fields generated by the compiler (for both auto-properties and anonymous-types)
+			return type.GetFields(PropertiesOrFieldOfClass).Where(x => !x.Name.StartsWith('<'));
```
[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate/Util/TypeExtensions.cs]: It seems the answer is in the PR's title: "use more lightweight method finding user's fields". The comment needs an update then.

File: src/NHibernate/Proxy/AbstractProxyFactory.cs [hunk 2]
```diff
@@ -41,9 +42,14 @@ protected bool IsClassProxy
 			GetIdentifierMethod = getIdentifierMethod;
 			SetIdentifierMethod = setIdentifierMethod;
 			ComponentIdType = componentIdType;
-			OverridesEquals = ReflectHelper.OverridesEquals(persistentClass);
+			OverridesEquals = ReflectHelper.OverridesEquals(persistentClass) &&
+			                  HasUserDeclaredFields(persistentClass);
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Proxy/AbstractProxyFactory.cs]: Sure

File: src/NHibernate/Util/TypeExtensions.cs [hunk 3]
```diff
@@ -48,5 +52,28 @@ internal static System.Type UnwrapIfNullable(this System.Type type)
 
 			return type;
 		}
+
+		public static IEnumerable<MemberInfo> GetFieldsOfHierarchy(this System.Type type)
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Util/TypeExtensions.cs]: R# did this. I had this internal. It should be internal indeed.

File: src/NHibernate/Util/TypeExtensions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Util/TypeExtensions.cs]: The comment was there :) I don't know who wrote it.  I just move the method from one TypeExtensions to another (in different namespace, as it now consumed not only by ByCode, but proxies as well)

File: src/NHibernate/Util/TypeExtensions.cs [hunk 4]
```diff
@@ -48,5 +52,28 @@ internal static System.Type UnwrapIfNullable(this System.Type type)
...
+					{
+						yield return fieldInfo;
+					}
+
+					analyzing = analyzing.BaseType;
+				}
+			}
+		}
+
+		public static IEnumerable<FieldInfo> GetUserDeclaredFields(System.Type type)
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Util/TypeExtensions.cs]: It was like this. Should be internal as well.

File: src/NHibernate/Util/TypeExtensions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate/Util/TypeExtensions.cs]: Now I have checked all that a bit more, with [this fiddle](https://dotnetfiddle.net/LCPaxG) for C# and [this one](https://dotnetfiddle.net/Lt7DZx) for VB.

So the trick is, anonymous type members are generated as properties with backing fields (at least in C#). The backing field name starts with `<`, but is not flagged with `CompilerGeneratedAttribute`. So indeed for C#, the code comment is valid. 

Still this falls short for the VB case. We do support entities coded in VB, don't we? For the VB case, anonymous fields could be excluded by filtering names starting with `$`. But about auto-properties fields, their naming is too usual (just an `_` prefix). Only the check on `CompilerGeneratedAttribute` seems reliable for them.

Then there is the F# case, and after all, the case of all the other languages available under .Net.

**Comment by fredericDelaporte** [comment]
> This PR does not have any timeout issues

I had already acknowledge it in the comment on the fix I had already done:

>Stupid bug indeed. Of course removed by Alexander's PR by the way, but I want first to check the user field detection algorithm.

**Pull request merged**